clippy_lints/src/transmute/eager_int_transmute.rs

```diff
@@ -0,0 +1,39 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::{path_to_local, path_to_local_id};
+use rustc_hir::{Expr, ExprKind, Node};
+use rustc_lint::LateContext;
+
+use super::EAGER_INT_TRANSMUTE;
+
+fn peel_parent_unsafe_blocks<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> Option<&'tcx Expr<'tcx>> {
+    for (_, parent) in cx.tcx.hir().parent_iter(expr.hir_id) {
+        match parent {
+            Node::Block(_) => {},
+            Node::Expr(e) if let ExprKind::Block(..) = e.kind => {},
+            Node::Expr(e) => return Some(e),
+            _ => break,
+        }
+    }
+    None
+}
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, transmutable: &'tcx Expr<'tcx>) -> bool {
+    if let Some(then_some_call) = peel_parent_unsafe_blocks(cx, expr)
+        && let ExprKind::MethodCall(path, receiver, ..) = then_some_call.kind
+        && cx.typeck_results().expr_ty(receiver).is_bool()
+        && path.ident.name == sym!(then_some)
+        && let ExprKind::Binary(_, lhs, _) = receiver.kind
+        && let Some(local_id) = path_to_local(transmutable)
+        && path_to_local_id(lhs, local_id)
+    {
+        span_lint_and_help(
+            cx,
+            EAGER_INT_TRANSMUTE,
+            expr.span,
+            "this transmute is evaluated eagerly, even if the condition is false",
+            Some(path.ident.span),
+            "consider using `then` instead, which evaluates it only if the condition is true",
```

Comment:
(Sorry if this is wrong)
I think suggesting this symbol requires MSRV-gating because [it](https://doc.rust-lang.org/stable/std/primitive.bool.html#method.then) is stabilized in 1.50?